### ISSUE TITLE
An alternative Search API proposals

### ISSUE THREAD

**Issue by mcupak** [ga4gh-discovery/ga4gh-case-discovery#75]
There are 2 different Search API proposals documented through various threads and documents - the one in this repository, and the one proposed by Marc Fiume, as shown in the screenshot below:

![image](https://user-images.githubusercontent.com/1255598/54691912-82e78080-4afa-11e9-9910-3396d0eb8fa4.png)

We should figure out how to align them and converge on a single API offering the benefits of both.

**Comment by Relequestual** [comment]
This is indeed very different. I thought that what was demonstrated in this repository currently, and released as 1.0.0 RC5, was the structure that most aligned with the needs, and was a view of general consensus from the currently interested groups. (As in, what's in the repository is not a proposal... The group has gone round and round on this, and now there are three demonstrations that show the structure works.)

What you've included in your screenshot here is a totally different approach.
Can you express what you feel the benefits are over the existing agreed on structures?

It looks like you're trying to build something that can be directly and easily translated into SQL.

I've expressed a few times my missgivings of this approach:
- There are existing solutions to this problem
- This is not the problem that the Search API is trying to solve

If this were the problem we were trying to solve, we should simply use existing solutions.

I outlined the reasoning in my initial rebutal proposal to Harindras straw man proposal: https://gist.github.com/Relequestual/65c0446944519a66f8562d02b3cb4c86

I'll include the requirements section here, as this is what I see as the key goals and considerations for the Search API.

> Requirements
> 
> _A standard workflow, method, and format, which can facilitate genomic data discoverability and sharing on a federated global network scale._
> 
> Because of the heterogeneity of the data involved, from patient phenotypes to variants, the API needs to be dynamic in nature, being able to cater for various types of data in any combination. The challenge this presents has been felt while developing the MME API, and the initial draft for a second version of MME is the foundation for this proposal.
> 
> A few key requirements, but not a complete set:
> 
> - Discoverability of patient data
> - Discoverability of variant data
> - Many possible search constraints
> - Many possible response data types
> - Flexibility on data type support
> - Low barrier to entry
> - Low cost to maintain

I feel this is a bigger discussion than we can have if we want to actually release RC6.

I see this issue as a suggested re-scoping of the Search API, and as such I'm removing it from the RC6 milestone.

**Comment by kemp-google** [comment]
For what it's worth:
* I do think having a standard query language is valuable (eg, either SQL or something directly translatable to it, or something else that's standard).  What is proposed in the specification right now seems very limited to me and will no doubt grow over time
* If there is truly a feeling that there are existing technologies that cover this problem space, I think it's worth documenting exactly why they don't satisfy the requirement list Ben has above (i.e., alternatives considered).  If this already exists and I just don't know about it, please point me to it!

**Comment by Relequestual** [comment]
There was a lot of very early days discussion around this where we carried out the excercise you describe. I spent several weeks on this.

One issue is, we were told we need to NOT tie anything down to specific implementations (for example, SQL), and GA4GH is a specifications game, as such.

Of course, whatever is specified needs to be easily translatable into SQL. I do not debate that being able to specify different query operations that you'd be able to in SQL.

Currently the specification is limited in terms of those operators, which should be expanded. We decided to start with, it would be kept super simple, to prove the utility of the framework.

I'm happy to document other possible approaches previously considered, and why I feel they are not suitable for the use cases presented. This needs to be documented (in the repo, in a new markdown file), which can be pointed to. It's my own fault for not having done that last time. I didn't expect to be in the position to need to justify everything again before there were 1.0.0 implementations up and running.

**New issue needs to be opened** to document other proposed approaches by using existing tech, and why they would fail to meet the requirements I see for this project. Happy to action that.

**Comment by Relequestual** [comment]
Also, it's worth remembering that the GA4GH PRC were happy with the framework concept. As a framework, it may have been approved, but there were concerns about how components would be maintainted and updated, and documentation issues to be resolved.

What I'm saying is, others have seen the structure as it is now, and agree that it's what's required.
I'm not saying there shouldn't be clear documentation for the reasoning behind decisions made.

**Comment by ljdursi** [comment]
At CanDIG, we're starting to work on the backend engine for this sort of search functionality - it's a requirement for us, and for the CINECA project we're part of along with many of our EU partners.  Any such backend will require farming out the query to multiple services, in parallel or sequentially, and combining the results. (e.g., the prototype involves clinical data from a FHIR service and germline variant data in a simple data store).   On top of that backend, we will try to implement whatever API this group decides.

We really liked the broad approach to the problem that had been outlined at the Basel meeting, and on the data side, the introduction of bioschemas begins to answer the important question of "where do the data models come from" for us.

We definitely agree with the requirements posted above, particularly towards the end:
```
* Many possible search constraints
* Many possible response data types
* Flexibility on data type support
* Low barrier to entry
* Low cost to maintain
```
For CINECA, we are required to implement this service multiple times on top of many sites' different home-grown clinical and variants (for instance) services/data stores by the end of the year, so low barrier to entry/cost to maintain is crucial!

To us that does mean something like queries that are easily translated into some widely available query language, whether it's GraphQL or SQL, so that we can easily partition the query and farm those sub-queries out to the relevant services, or (even better) use existing tools that will do that for us.

While I'm a big proponent of DSLs in a lot of contexts, I'm more cautious here - custom-made query languages don't generally fare very well (I've never met a single person who's ever used oData, for instance), and we just don't have the personnel to implement a custom (and presumably evolving) query language 6-8 times with different partitioning logic for the different sites within the next 8 months.

**Comment by Relequestual** [comment]
Hey @ljdursi Thanks for your comments. I've yet to review the use case collection document from CanDIG, but I'm appreciative of it being filed and sent our way.
I agree, it's important to have something that is easily translatable into whatever query language the end system wants to use, but we shouldn't take a specific one (like SQL), and create an abstraction layer ontop of that as the format. That would be like an ORM query abstraction library format, of which there are MANY, and they do not fulfil the requirements of this project.

I'd be keen to hear your feedback on the technical level of the current version of the Search API. Please email me or the Search API google group if you'd rather, or simply if you want to chat over it. (Let me know if you don't have my email. Following you already on twitter.)

The component based architecure define in this API is the result of many many hours of work, initially by myself and Orion (who I believe you know). I'm still confident I can demonstrate the utility and how it fulfils the use cases and requirements laid out.

**Comment by mcupak** [comment]
@Relequestual Regarding the approaches previously considered, what technologies were evaluated (e.g. SQL, GraphQL)?

**Comment by Relequestual** [comment]
We looked at potentially using something like elastic search or solr, the mandate is to not limit use to a specific tech stack (which that would do).

I agree we should have something that's easily translatable into SQL, but the issues driver projects face are a layer above acutally constructing a query.

At the risk of [repeating myself](https://github.com/ga4gh-discovery/ga4gh-discovery-search/issues/75#issuecomment-474864609), what's in the repository is not a proposal. It's a the culmination of over a year of work and years of experience running into problems as part of the MME.

@mcupak I feel a better approach here would be to explain what you feel the current version of the Search API doesn't facilitate. We can then be pragmatic about changes, rather than a sudden presentation of a radically different approach (which this issue proposes).

The Search API as is, fulfills several driver project requirements, and we must ensure any changes also reflects the needs of driver projects. I haven't seen any driver project raise concerns of limitiations on the Search API. In fact, the opposite, current implementers have said it translates to SQL quite well.

**Comment by fschiettecatte** [comment]
I would second @Relequestual's comments, drawing a spec to suit an implementation (SQL for example) would incur 'technical debt' that would be repaid in spades in the future. The aim of the spec is to represent a search, and let individual sites implement it how they best see fit on top of or alongside whatever search/storage systems they have.

VariantMatcher uses MySQL for storage and SOLR for search, while GeneMatcher uses MySQL for both. VariantMatcher supports the current spec of DiscoverySearch and GeneMatcher supports MME v1.x. In both cases the systems were built before either DS or MME were a 'thing', and in both cases a layer of code was be implemented to translate the  DS / MME search & results. In both cases this was easy because DS / MME are well designed abstractions. At its core the DS spec is 'just' a boolean search.

**Comment by ljdursi** [comment]
I sympathize with the wariness, but the idea that relying on a home-made and hand-implemented query language is _not_ technical debt while being able to easily use a widely implemented query language in the back end _is_ technical debt is strange to me.

The idea that a particular choice easily allowing (not requiring, allowing) use of a standard query language - SQL or GraphQL - as an interface to stores (e.g., [Solr SQL](https://lucene.apache.org/solr/guide/6_6/parallel-sql-interface.html)) is a bug not a feature is equally odd to me, and I must be missing something.  Is the concern that SQL is going to go away soon?  Or that there are queries we'll need that are unimplementable in SQL?  There may be back end stores somewhere which still lack decent enough SQL frontends, but then the situation one is in - having to hand-code a translation between the query and the store - is no worse than the situation one was in before.

Every technical decision is lock-in to some extent; the question here is whether a decision is vendor lock-in to something where you are both the only vendor and only customer.

**Comment by Relequestual** [comment]
@ljdursi If the problem here was just allowing the federation of queries to SQL capable backend systems, based on a set of standard schemas, then there would be no need for this project.

It is expected that the Search API structure will be translated into SQL, and we should make that easy.
During the early discussion, I even suggested the `logic` object, could follow a structure used in an existing fully featured SQL abstraction format used by popular ORMs.

The issues the Search API looks to solve are at a higher level than the query language that interacts with the actual data store.

The currently demonstrated presto implementation relies on systems to return the fields they support, and then allows you to construct a query based on those fields. This can only scale so far. Presenting the user with 30 different fields for the same thing would make it untenable.

The Search API structure allows the querying server to not NEED to know what fields or even sets of fields the endpoints they want to query support, so federation can be as broad as possible, and the results as useful as possible.

The issues at play are mostly not technical, but sociological.

For example, some projects involved in MME have a much more limited ability to adapt to change. A major version bump means you cut off half the network for months while they find someone to implement the changes for them.
Now, in a world where your schemas are evolving at a reasonable pace with semantic versioning, you've potentially perminantly locked out any number of projects from being able to participate in the federated query network.


The structure should be easily translatable into SQL, yes, but it should not BE SQL. No query language is designed to solve these types of sociological issues, at least none that I'm aware of, and believed me I've looked.

**Comment by kemp-google** [comment]
> @ljdursi If the problem here was just allowing the federation of queries to SQL capable backend systems, based on a set of standard schemas, then there would be no need for this project.

I do not think the use of a standard query language (like SQL) and the underlying data model are tied together as closely as you suggest.

What I think Marc was proposing is in fact not really any different than what you have proposed, except that instead of inventing a new method to construct queries, it proposes using an existing standard.

The query and the model are separable: we *should* define common data models - as you say, we can't have "30 different fields" for the same thing.  But this doesn't mean we can't *query* them with something standard like SQL.

    `SELECT alternate FROM org.ga4gh.variants AS v WHERE v.start > 1000 AND v.end < 2000`

Implementations can take this and perform their search however they like: either by mapping these fields/virtual-tables to real tables/fields or perhaps by querying data stores that aren't even SQL based (as in the Presto implementation presented by Marc's team - they are querying other GA4GH servers, not just SQL data stores). 

> It is expected that the Search API structure will be translated into SQL, and we should make that easy.

I think the question that keeps being surfaced is given that, why not just start with SQL?  It doesn't have any determining effect on implementations (other than that there are existing implementations of SQL parsers that could be leveraged).

> The currently demonstrated presto implementation relies on systems to return the fields they support, and then allows you to construct a query based on those fields. This can only scale so far. Presenting the user with 30 different fields for the same thing would make it untenable.

I think the `/fields` endpoint above is maybe causing some undue confusion: I believe (@mcupak can correct me) that the intent was to allow this so that additional data could be discovered, but the intent would still be that implementations would support well known ga4gh models.

**Comment by Relequestual** [comment]
As an immediate reaction, I'd be a little worried if we started passing around SQL, that people would send it directly to the database, which obviously creates huge security risks. While it's true that all paramaters should be parameterized anyway, passing a fully fledged SQL query certinly increases the chance that someone will assume it's safe.

Additionally, I'm concerned that having to use an SQL parser would massivly increase the barrier to entry. It should be as easy to implement and be compliant as Beacon, but allow for extensibility way beyond the complexity of MME.

If by SQL parser you mean something like https://github.com/mozilla/moz-sql-parser, then yeah I suggested the `logic` element could use a similar structure. (or https://github.com/taozhi8833998/node-sql-parser or https://github.com/godmodelabs/flora-sql-parser)

If it has to be parsed into a structure to then be recreated as SQL or handed off to some other function (like to access files), then why not start with that structure and miss out that step?

Regardless, I'm going to give your comments some further thought.

**Comment by rishidev** [comment]
I have been having some confusion on calls trying to follow discussions and I think this thread crystallizes the issue.  The first step to resolving this is I think to clarify terminology. The Discovery Search API is and should be what is based in this repository. It has been worked through with various Driver Projects - MME, EJP-RD, HCA, ENA/EVA/EGA.

 The second set should probably be referred to as something different - it seems there are non-mergable approaches in philosophy with the main Search API. There is no problem with us having two separate products issued, as long as they are not non-compatible. It seems there is a demand for this. Should this actually be spun off as a separate, more implementation focused, API, with a translational layer to the Search API? It would really make discussions clearer and more transparent if this were spun out into a separate repo.

For the existing Search API - As has been pointed out the issues with this at the GA4GH Steering Committee were ones of governance and not enough representation of component types in the demonstration. I believe the former has been dealt with via schema blocks with details still to be ironed out and the later.

**Comment by mcupak** [comment]
Many good points here.

@rishidev I think you identified the problem correctly in that there are 2 approaches right now and the message going outside of this group is conflicting because of that. Both approaches are trying to solve a very similar problem though and ultimately enable the same use cases so I'd prefer to align as part of a single standard to avoid further confusion. I'd consider 2 separate products as a last resort if we diverge further.

As for the other points, I agree with basically everything @ljdursi and @kemp-google said.

If we can use an existing query language, that's absolutely better than inventing a new one. The important thing is the size of the ecosystem, really. There are many tools across all the possible frameworks and languages that make it easy to produce, and importantly, consume SQL. Most storage systems (incl. NoSQL stores) provide an SQL abstraction layer - if the main thing we're looking for is being able to map to many different backends, it's really hard to find something better than SQL. We'd never be able to reach that level of support and convenience of use with our own language. 

The fact that the data model and the query language are, in fact, independent, is very important. Case in point, SQL is probably one of the most successful standards in the world and it doesn’t come with a data model. Neither do other *QL.

Based on what I’ve seen in practice so far, I also think the query language part is a much easier problem to solve than the data model part. Harmonizing data is hard. I’d love it if everyone could just implement our standard data model, but that’s not going to be the case in the real world. People already have a lot of data collected in various schemas and if they want to share it and enable search on it, we shouldn’t prevent them from doing that. Depending on a fixed data model makes this initial jump very big and expensive.

I am absolutely for Schemablocks as the missing piece here, we should provide a standard data model, because having that unlocks a lot of value for federated queries, but we do get substantial value without it as well. Realistically, it’s going to be just one of many data models people use. We can enable people to use their own models through the Schemablocks catalog, if it’s sufficiently flexible and lightweight, but then we also have use cases like NIH, where they have many datasets with different schemas, so why not just have the API to tell you directly what you can query on (`/fields` in the screenshot, or `DESCRIBE` in SQL)?

The main argument I’m hearing for developing our own query language is the fact that SQL is complex and we only need something much simpler. Firstly, I think people overestimate how hard SQL is to implement. [Presto's own SQL grammar](https://github.com/prestodb/presto/blob/master/presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4) is only a few hundred lines of code, plus there are many open source parsers and other tools out there. Secondly, we might need a very simple API now, but that API is going to grow more complex over time, and I worry we’ll be stuck with a complex API without its ecosystem of tools, essentially reinventing SQL.

**Comment by asbjornu** [comment]
@mcupak, first; 700 lines of code **is not simple**. Second, I've never seen a database abstraction layer that doesn't in one way or another leak features of the underlying database up into the client. Especially with Presto claiming support for both SQL and non-SQL databases, I'm having a very hard time to believe there's 100% feature parity and compatibility across all database backends.

This means clients are going to perform queries that make them dependent on the database backend, forcing the question: Why did we introduce this abstraction layer in the first place?

SQL also gives a client the largest API surface that is possible to imagine. It gives clients the possibility to deeply couple itself to all of the internal details of the server, completely unfiltered, unchecked and unbounded. If this is what you want, why not just give the clients an SQL connection string and let them perform the SQL statements directly against the database, natively?

That being said, I do agree developing your own query language is also be a bad idea. There are many in existence that you can pick up and use, which give you much more control over the API surface you give to the clients and how to deal with evolving that API surface over time as the server and the requirements it operates under changes. [OData](https://www.odata.org/), [SPARQL](https://www.w3.org/TR/rdf-sparql-query/) and [GraphQL](https://graphql.org/), for instance.

**Comment by ljdursi** [comment]
Re: other query languages -  I was actively advocating for GraphQL for exactly this purpose, until we started working with it more closely.  There are definitely places it fits nicely - exposing connections between graph-like data like complex FHIR records - but it's only barely a query language in the sense we'd like here (not including basic things like filtering as part of the standard syntax) and for that reason we're probably moving to PRESTO internally for our cross-API queries internally even including FHIR.  (We're actually going to write up a blog post or something talking about our experiences around comparing them in a simple test bed: https://github.com/CanDIG/presto-graphql-testbed).

Similarly, SPARQL works really well for RDF, but that's a pretty specific use case.

For OData, I've always really liked the idea of it, particularly the very ORM-y approach to relationships - but I've yet to meet anyone who's actually used it; the tools supporting it are pretty few and far between.  And maybe I'm wrong, but I've always just thought of it as a particular mapping from REST to SQL - e.g. examples on the website:

`GET https://services.odata.org/v4/TripPinServiceRW/People?$top=2 &amp; $select=FirstName, LastName &amp; $filter=Trips/any(d:d/Budget gt 3000) `

that's pretty transparently something like
```
SELECT
  Firstname,
  Lastname
FROM
  People
  INNER JOIN Trips ON People.PersonID = Trips.PersonID
WHERE
  Trips.Budget > 3000
GROUP BY
  People.PersonID
LIMIT
  2;
```

(or something, my SQL is pretty rusty).

**Comment by mcupak** [comment]
Closing as per recent decisions on calls and emails. This repository will evolve to contain case discovery API, and search will be developed as part of [ga4gh-search](https://github.com/ga4gh-discovery/ga4gh-search).
